docs/halibut-protocol.md

```diff
@@ -29,15 +30,46 @@ The final stage can be entered by either the Client or Service terminating the T
 
 ## Listening Service protocol data exchange
 
-When the **Service** is in  **Listening** mode, the **Client** identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.
+In listening mode the *Client* will keep and pool connections for subsequent RPC calls. In this mode the "pooled connection" is tested by sending and recieving control messages before sending the Request message. If the control messages can not be exchanaged a new connection is made.
+
+1. *Identification:*
+    1. **Client** identifies itself as `MX-CLIENT 1.0`
+    2. **Service** identifies itself as `MX-SERVER 1.0`
+2. *Message exchange:*
+    1. The **Client** sends a `Request` message containing the method to execute, its parameters and 0 or more data streams.
+    2. The **Service** executes the method and sends the result in a `Response` message containing the result and 0 or more data streams.
+    3. The **Client** places the TCP connection into the pool
+    4. The **Service** waits for a control message.
+3. *Connection re-use: When the Client wants to make a new RPC call:*
+    1. A connection is removed from the pool and tested by:
+        1. The Client sends a `NEXT` control message.
+        2. The Service responds witha `PROCEED` control message.
+    2. If the client is unable to send `NEXT` or recieve the `PROCEED` control message, the TCP connection is clossed and a new one created following the process from step `1.Identification`.
+    3. If the client recieves `PROCEED` the RPC call is made with this TCP connection starting at step `2.Message exchange`.
+
+
 
 ![Listening client protocol data exchange](images/listeningprotocoldata.png)
 
 ## Polling Service protocol data exchange
 
-When the **Service** is in  **Polling** mode, the **Service** identifies itself as `MX-SUBSCRIBER 1.0` while the `Client` identifies itself as `MX-SERVER 1.0`.
+In Polling mode the **Service** initiates the connection and does a long poll to the **Client** for Requests. In this mode only a single TCP connection exists at any one time, unlike in listening mode where connections are created as required. This means that in polling mode, only one RPC call can be made at a time.
+
+To make RPC calls the steps are:
+
+1. *Identification:*
+    1. **Service** identifies itself as `MX-SUBSCRIBER 1.0`
+    2. **Client** identifies itself as `MX-SERVER 1.0`
+2. *Message exchange:*
+    1. The Client sends a `Request` message containing the method to execute, its parameters and 0 or more data streams.
+    2. The Service executes the method and sends the result in a `Response` message containing the result and 0 or more data streams.
+3. *Connection re-use:*
+    1. The **Service** sends a `NEXT` control message.
+    2. The **Client** responds with a `PROCEED` control message.
+    3. The **Service** Waits for a **Request** message. If waiting times out the connection is closed and recreated starting again from step `1.Identification`.
```

Comment:
Is this correct? I thought if there was no message and waiting timed out then the Client sent a null request and the process went around again, not that the connection gets recreated?

Reply:
The bit at the end about No-ops being sent I think covers that case.

Reply:
Ah right, I'm still finding it a bit hard to follow as this makes it sound like the normal case will be that the connection drops and gets created. It might be worth adding both the cases here as sub-points so that the normal case of the connection being held open is clear.

Reply:
I agree I think this could be improved, but I think I will merge this as is since it is better and we can come back to it when we have time and focus.